File: ruby/day_03.rb
def parse_input(fname)
  File.readlines(fname).collect { |l| l.chomp.to_i(2) }
end

def count_ones(words, bit)
  words.count { |w| (w & bit).positive? }
end

def select_by(words, word_size, cond)
  hi_bit = 1 << (word_size - 1)
  selected = words.dup
  (0...word_size).each do |i|
    bit = hi_bit >> i
    ones = count_ones(selected, bit)
    if cond.call(ones, selected.length)
      selected.keep_if { |r| (r & bit).positive? }
    else
      selected.keep_if { |r| (r & bit).zero? }
    end
    break unless selected.length > 1
  end
  selected[0]
end

def part_one(words, word_size)
  half = words.length / 2
  hi_bit = 1 << (word_size - 1)
  gamma = 0
  epsilon = 1
  (0...word_size).each do |i|
    bit = hi_bit >> i
    c = count_ones(words, bit)
    if c > half
      gamma |= bit
    else
      epsilon |= bit
    end
  end
  gamma * epsilon
end

def part_two(words, word_size)
  most_common = ->(ones, count) { ones >= (count + 1) / 2 }
  least_common = ->(ones, count) { ones < (count + 1) / 2 }
  oxy = select_by(words, word_size, most_common)
  co2 = select_by(words, word_size, least_common)
  oxy * co2
end

test_words = %w[00100 11110 10110 10111 10101 01111 00111 11100 10000 11001 00010 01010].map { |n| n.to_i(2) }
puts "#{part_one(test_words, 5)}, #{part_two(test_words, 5)}"

words = parse_input('../inputs/day_03.txt')
puts "#{part_one(words, 12)}, #{part_two(words, 12)}"

File: clojure/day_03.clj
(ns day_03
  (:require [clojure.string :as str]))

(def test-report
  (map #(Long/parseLong % 2)
       '("00100" "11110" "10110" "10111" 
         "10101" "01111" "00111" "11100" 
         "10000" "11001" "00010" "01010")))

(def diagnostic-report
  (->> (slurp "../inputs/day_03.txt")
       (str/split-lines)
       (map #(Long/parseLong % 2))))

(defn select-zeros
  [numbers bit]
  (let [bit-mask (bit-shift-left 1 bit)]
    (filter #(zero? (bit-and % bit-mask)) numbers)))

(defn select-ones
  [numbers bit]
  (let [bit-mask (bit-shift-left 1 bit)]
    (filter #(pos? (bit-and % bit-mask)) numbers)))

(defn count-ones
  [numbers bit]
  (count (select-ones numbers bit)))

(defn select-by-bit-value
  [numbers bit value]
  (cond
    (zero? value) (select-zeros numbers bit)
    (pos? value) (select-ones numbers bit)))

(defn most-frequent-bit
  [numbers bit]
  (let [half (/ (count numbers) 2)]
    (if (>= (count-ones numbers bit) half) 1 0)))

(defn least-frequent-bit
  [numbers bit]
  (let [half (/ (count numbers) 2)]
    (if (< (count-ones numbers bit) half) 1 0)))

(defn bits-to-number
  [bits]
  (reduce (fn [num bit] (+ (* num 2) bit)) 0 bits))

(defn gamma
  [numbers word-size]
  (bits-to-number
   (map (partial most-frequent-bit numbers)
        (range (- word-size 1) -1 -1))))

(defn epsilon
  [numbers word-size]
  (bits-to-number
   (map (partial least-frequent-bit numbers)
        (range (- word-size 1) -1 -1))))

(defn oxigen-rating
  [numbers bit]
  (if (= 1 (count numbers))
    (first numbers)
    (oxigen-rating
     (select-by-bit-value numbers (dec bit) (most-frequent-bit numbers (dec bit)))
     (dec bit))))

(defn co2-rating
  [numbers bit]
  (if (= 1 (count numbers))
    (first numbers)
    (co2-rating
     (select-by-bit-value numbers (dec bit) (least-frequent-bit numbers (dec bit)))
     (dec bit))))

(defn part-1
  [report word-size]
  (* (gamma report word-size) 
     (epsilon report word-size)))

(defn part-2
  [report word-size]
  (* (oxigen-rating report word-size)
     (co2-rating report word-size)))

(part-1 test-report 5)
(part-1 diagnostic-report 12)

(part-2 test-report 5)
(part-2 diagnostic-report 12)
